dist/datalayer.js

```diff
@@ -207,95 +208,161 @@ governing permissions and limitations under the License.
                 'because it does not have a valid format: ' +
-                JSON.stringify(item);
+                JSON.stringify(item.getConfig());
             console.error(message);
             return;
         }
 
         var typeProcessors = {
             data: function(item) {
                 that._updateState(item);
-                that._triggerListeners(item.getConfig(), events.CHANGE);
+                that._triggerListeners(item);
             },
             event: function(item) {
-                that._triggerListeners(item);
                 if (item.getConfig().data) {
                     that._updateState(item);
-                    that._triggerListeners(item.getConfig(), events.CHANGE);
                 }
+                that._triggerListeners(item);
             },
             listenerOn: function(item) {
-                that._registerListener(item.getConfig());
+                that._processListenerOn(item);
             },
             listenerOff: function(item) {
-                that._unregisterListener(item.getConfig());
+                that._unregisterListener(item);
             }
         };
 
         typeProcessors[item.getType()](item);
     };
 
     /**
-     * Updates the state with the passed data configuration.
+     * Processes the item of type: listener on.
      *
-     * @param {DataConfig} item The data configuration.
+     * @param {DataLayer.Item} listener The listener.
      * @private
      */
-    DataLayer.Manager.prototype._updateState = function(item) {
-        DataLayer.utils.deepMerge(this._state, item.data);
+    DataLayer.Manager.prototype._processListenerOn = function(listener) {
+        var scope = listener.getConfig().scope;
+        if (!scope) {
+            scope = listenerScope.FUTURE;
+        }
+        switch (scope) {
+            case listenerScope.PAST:
+                // trigger the handler for all the previous items
+                this._triggerListenerOnPreviousItems(listener);
+                break;
+            case listenerScope.FUTURE:
+                // register the listener
+                this._registerListener(listener);
+                break;
+            case listenerScope.ALL:
+                // trigger the handler for all the previous items
+                this._triggerListenerOnPreviousItems(listener);
+                // register the listener
+                this._registerListener(listener);
+                break;
+            default:
+                console.error('The listener does not have a valid scope: ' + scope);
+        }
     };
 
-    DataLayer.Manager.prototype._triggerListeners = function(item, event) {
+    /**
+     * Calls all the handlers matching the specified item.
+     *
+     * @param {DataLayer.Item} item The item.
+     * @private
+     */
+    DataLayer.Manager.prototype._triggerListeners = function(item) {
         var that = this;
+        that._listeners.forEach(function(listenerConfig) {
```

Comment:
Instead of looping over all listeners, do you think it would make more sense to structure these by key:
```
._listeners = {
  "datalayer:change": [],
  "datalayer:event" : [],
  "datalayer:ready" : [],
  "<custom>": []
};
```

Reply:
We could then pre-switch the logic and `._triggerListeners` could just accept the event to trigger?

```
._triggerListeners(key) {
  switch (key)
    case: "datalayer:change"
    case: "datalayer:event"
    case: "datalayer:ready"
    case: <custom>
}
```

Reply:
yes we could do that

Reply:
I will change the logic as suggested.